File: Design-Patterns-Using-Java/Proxy/ProxyLazy.java
interface Subject {
    void request();
}

class RealSubject implements Subject {
    @Override
    public void request() {
        System.out.println("Concrete Subject Request Method");
    }
}

class Proxy implements Subject {
    private RealSubject realSubject;

    @Override
    public void request() {
        if (realSubject == null) {
            realSubject = new RealSubject(); // Lazy Initialization
        }
        realSubject.request();
    }
}

public class ProxyLazy {
    public static void main(String[] args) {
        Proxy proxy = new Proxy();
        proxy.request();
    }
}
/* 
Concrete Subject Request Method
*/

File: Design-Patterns-Using-CSharp/Proxy/ProxyLazy.cs
using System;

// Subject Interface
interface ISubject
{
    void Request();
}

// RealSubject Class
class RealSubject : ISubject
{
    public void Request()
    {
        Console.WriteLine("Concrete Subject Request Method");
    }
}

// Proxy Class
class Proxy : ISubject
{
    private RealSubject realSubject;

    public void Request()
    {
        if (realSubject == null)
        {
            realSubject = new RealSubject(); // Lazy Initialization
        }
        realSubject.Request();
    }
}

// Client code
class ProxyLazy
{
    static void Main(string[] args)
    {
        Proxy proxy = new Proxy();
        proxy.Request();
    }
}

/*
Concrete Subject Request Method
*/